File: Dashboard/backend/lambdaProcessor.js
// ============================================================
// lambdaProcessor.js — AWS Lambda Function Simulator
// ============================================================
// Subscribes to the Kinesis stream and processes each record:
// transforms raw OpenWeatherMap data into analytics-ready
// format with derived metrics, then stores in DataStore.
// ============================================================

class LambdaProcessor {
    constructor(kinesisStream, dataStore) {
        this.functionName = 'WeatherDataProcessor';
        this.kinesisStream = kinesisStream;
        this.dataStore = dataStore;
        this.invocationCount = 0;
        this.errorCount = 0;
        this.lastInvocation = null;
        this.status = 'ACTIVE';
        this.averageDuration = 0;
        this._totalDuration = 0;

        // Subscribe to Kinesis stream events (event source mapping)
        this.kinesisStream.on('record', (record) => this._handleRecord(record));

        console.log(`[Lambda] Function "${this.functionName}" deployed and connected to Kinesis stream`);
    }

    /**
     * Process a single Kinesis record (the Lambda handler).
     */
    _handleRecord(record) {
        const startTime = Date.now();
        this.invocationCount++;

        try {
            const rawData = record.data;

            // Transform raw weather data into analytics-ready format
            const processed = this._transformWeatherData(rawData);

            // Store processed data
            this.dataStore.put(processed.city, processed);

            const duration = Date.now() - startTime;
            this._totalDuration += duration;
            this.averageDuration = Math.round(this._totalDuration / this.invocationCount);
            this.lastInvocation = new Date().toISOString();

            console.log(`[Lambda] ✓ Processed: ${processed.city} | Temp: ${processed.temperature}°C | Duration: ${duration}ms`);

            return { statusCode: 200, body: processed };
        } catch (err) {
            this.errorCount++;
            console.error(`[Lambda] ✗ Error processing record:`, err.message);
            return { statusCode: 500, error: err.message };
        }
    }

    /**
     * Transform raw OpenWeatherMap API data into dashboard-ready format.
     */
    _transformWeatherData(raw) {
        const temp = raw.main?.temp ?? 0;
        const feelsLike = raw.main?.feels_like ?? 0;
        const humidity = raw.main?.humidity ?? 0;
        const windSpeed = raw.wind?.speed ?? 0;
        const windDeg = raw.wind?.deg ?? 0;

        return {
            // Identity
            city: raw.name || 'Unknown',
            country: raw.sys?.country || 'N/A',
            cityId: raw.id,
            coordinates: raw.coord || { lon: 0, lat: 0 },

            // Core metrics
            temperature: Math.round(temp * 10) / 10,
            feelsLike: Math.round(feelsLike * 10) / 10,
            tempMin: Math.round((raw.main?.temp_min ?? temp) * 10) / 10,
            tempMax: Math.round((raw.main?.temp_max ?? temp) * 10) / 10,
            humidity,
            pressure: raw.main?.pressure ?? 0,
            seaLevel: raw.main?.sea_level ?? null,
            groundLevel: raw.main?.grnd_level ?? null,

            // Wind
            windSpeed: Math.round(windSpeed * 10) / 10,
            windDirection: windDeg,
            windDirectionLabel: this._degToCompass(windDeg),
            windGust: raw.wind?.gust ? Math.round(raw.wind.gust * 10) / 10 : null,

            // Conditions
            weatherMain: raw.weather?.[0]?.main || 'N/A',
            weatherDescription: raw.weather?.[0]?.description || 'N/A',
            weatherIcon: raw.weather?.[0]?.icon || '01d',
            cloudiness: raw.clouds?.all ?? 0,
            visibility: raw.visibility ?? 10000,

            // Sun
            sunrise: raw.sys?.sunrise ? new Date(raw.sys.sunrise * 1000).toISOString() : null,
            sunset: raw.sys?.sunset ? new Date(raw.sys.sunset * 1000).toISOString() : null,

            // Derived metrics
            feelsLikeDelta: Math.round((temp - feelsLike) * 10) / 10,
            heatIndex: this._calculateHeatIndex(temp, humidity),
            windChill: this._calculateWindChill(temp, windSpeed),
            dewPoint: this._calculateDewPoint(temp, humidity),
            comfortIndex: this._calculateComfortIndex(temp, humidity, windSpeed),

            // Metadata
            timestamp: new Date().toISOString(),
            dataSource: 'OpenWeatherMap',
            rawTimestamp: raw.dt ? new Date(raw.dt * 1000).toISOString() : new Date().toISOString()
        };
    }

    /**
     * Convert wind degrees to compass direction.
     */
    _degToCompass(deg) {
        const directions = ['N', 'NNE', 'NE', 'ENE', 'E', 'ESE', 'SE', 'SSE',
            'S', 'SSW', 'SW', 'WSW', 'W', 'WNW', 'NW', 'NNW'];
        return directions[Math.round(deg / 22.5) % 16];
    }

    /**
     * Calculate heat index (feels hotter in humidity).
     */
    _calculateHeatIndex(tempC, humidity) {
        if (tempC < 27) return null;
        const t = tempC * 9 / 5 + 32; // to Fahrenheit
        const h = humidity;
        let hi = -42.379 + 2.04901523 * t + 10.14333127 * h
            - 0.22475541 * t * h - 0.00683783 * t * t
            - 0.05481717 * h * h + 0.00122874 * t * t * h
            + 0.00085282 * t * h * h - 0.00000199 * t * t * h * h;
        return Math.round((hi - 32) * 5 / 9 * 10) / 10; // back to Celsius
    }

    /**
     * Calculate wind chill (feels colder in wind).
     */
    _calculateWindChill(tempC, windSpeedMs) {
        const windKph = windSpeedMs * 3.6;
        if (tempC > 10 || windKph < 4.8) return null;
        const wc = 13.12 + 0.6215 * tempC - 11.37 * Math.pow(windKph, 0.16)
            + 0.3965 * tempC * Math.pow(windKph, 0.16);
        return Math.round(wc * 10) / 10;
    }

    /**
     * Calculate dew point.
     */
    _calculateDewPoint(tempC, humidity) {
        const a = 17.27, b = 237.7;
        const alpha = (a * tempC) / (b + tempC) + Math.log(humidity / 100);
        return Math.round((b * alpha) / (a - alpha) * 10) / 10;
    }

    /**
     * Calculate comfort index (0-100).
     */
    _calculateComfortIndex(tempC, humidity, windSpeedMs) {
        let score = 100;
        // Temperature penalty
        const idealTemp = 22;
        score -= Math.abs(tempC - idealTemp) * 3;
        // Humidity penalty
        const idealHumidity = 45;
        score -= Math.abs(humidity - idealHumidity) * 0.5;
        // High wind penalty
        if (windSpeedMs > 10) score -= (windSpeedMs - 10) * 2;
        return Math.max(0, Math.min(100, Math.round(score)));
    }

    /**
     * Get Lambda function stats (like CloudWatch metrics).
     */
    getStats() {
        return {
            functionName: this.functionName,
            status: this.status,
            invocationCount: this.invocationCount,
            errorCount: this.errorCount,
            successRate: this.invocationCount > 0
                ? Math.round((1 - this.errorCount / this.invocationCount) * 100) : 100,
            averageDurationMs: this.averageDuration,
            lastInvocation: this.lastInvocation
        };
    }
}

module.exports = LambdaProcessor;

File: Dashboard/backend/lambda_processor.py
# ============================================================
# lambda_processor.py — AWS Lambda Function Simulator
# ============================================================
# Simulates an AWS Lambda function that processes records from
# the Kinesis stream, transforms the weather data, and stores
# it in the DataStore.
# ============================================================

from datetime import datetime
from typing import Optional


class LambdaProcessor:
    """Simulates an AWS Lambda function processing Kinesis records."""
    
    def __init__(self, kinesis_stream, data_store):
        self.kinesis_stream = kinesis_stream
        self.data_store = data_store
        self.invocations = 0
        self.records_processed = 0
        self.errors = 0
        self.started_at = datetime.now().isoformat()
        
        # Subscribe to Kinesis stream
        self.kinesis_stream.on('record', self._process_record)
        
        print('[Lambda] ✓ Processor initialized and subscribed to Kinesis stream')
    
    def _process_record(self, record: dict) -> None:
        """Process a single record from Kinesis."""
        try:
            self.invocations += 1
            raw_data = record['data']
            city = record['partitionKey']
            
            # Transform the raw OpenWeatherMap data
            transformed = self._transform_weather_data(raw_data, city)
            
            # Store in DataStore
            self.data_store.put(city, transformed)
            
            self.records_processed += 1
            
            print(f'[Lambda] Processed {city}: {transformed["temperature"]}°C, {transformed["weatherDescription"]}')
            
        except Exception as e:
            self.errors += 1
            print(f'[Lambda] Processing error: {e}')
    
    def _transform_weather_data(self, raw: dict, city: str) -> dict:
        """Transform OpenWeatherMap API data into our format."""
        try:
            # Handle both API and mock data structures
            main = raw.get('main', {})
            weather = raw.get('weather', [{}])[0]
            wind = raw.get('wind', {})
            sys = raw.get('sys', {})
            coords = raw.get('coord', {})
            
            return {
                'city': city,
                'temperature': float(f"{main.get('temp', 0):.1f}"),
                'feelsLike': float(f"{main.get('feels_like', 0):.1f}"),
                'tempMin': float(f"{main.get('temp_min', 0):.1f}"),
                'tempMax': float(f"{main.get('temp_max', 0):.1f}"),
                'pressure': main.get('pressure', 0),
                'humidity': main.get('humidity', 0),
                'visibility': raw.get('visibility', 0),
                'uvIndex': raw.get('uvIndex', 0),
                'windSpeed': float(f"{wind.get('speed', 0):.1f}"),
                'windDeg': wind.get('deg', 0),
                'windGust': float(f"{wind.get('gust', 0):.1f}") if wind.get('gust') else None,
                'precipitation': raw.get('rain', {}).get('1h', 0) or raw.get('snow', {}).get('1h', 0) or 0,
                'cloudiness': raw.get('clouds', {}).get('all', 0),
                'weatherMain': weather.get('main', 'Unknown'),
                'weatherDescription': weather.get('description', 'unknown'),
                'weatherIcon': weather.get('icon', '01d'),
                'sunrise': sys.get('sunrise'),
                'sunset': sys.get('sunset'),
                'timezone': raw.get('timezone', 0),
                'latitude': coords.get('lat', 0),
                'longitude': coords.get('lon', 0),
                'timestamp': raw.get('dt') or int(datetime.now().timestamp()),
                'processedAt': datetime.now().isoformat()
            }
        except Exception as e:
            print(f'[Lambda] Transform error: {e}')
            raise
    
    def get_stats(self) -> dict:
        """Get Lambda processor statistics."""
        started_dt = datetime.fromisoformat(self.started_at)
        uptime_ms = int((datetime.now() - started_dt).total_seconds() * 1000)
        
        # Calculate success rate safely (fixed lint error by avoiding round())
        if self.invocations > 0:
            raw_rate = float(self.records_processed) / float(self.invocations) * 100.0
            success_rate = float(f"{raw_rate:.2f}")
        else:
            success_rate = 0.0
        
        return {
            'functionName': 'WeatherDataProcessor',
            'status': 'ACTIVE',
            'invocations': self.invocations,
            'recordsProcessed': self.records_processed,
            'errors': self.errors,
            'successRate': success_rate,
            'startedAt': self.started_at,
            'uptime': uptime_ms
        }
